src/main/java/tools/jackson/databind/jsontype/impl/AsPropertyTypeDeserializer.java

```diff
@@ -203,6 +206,17 @@ public Object deserializeTypedFromAny(JsonParser p, DeserializationContext ctxt)
         return deserializeTypedFromObject(p, ctxt);
     }
 
+    // [databind#1654]: Determine whether strict type-id handling should apply for this property.
+    private boolean isStrictTypeIdHandlingForProperty() {
+        if (!_strictTypeIdHandling) {
+            return false;
+        }
+        if (_property == null) {
+            return true;
+        }
+        JsonTypeInfo typeInfo = _property.getAnnotation(JsonTypeInfo.class);
```

Comment:
This is too expensive call to make during actual reading -- that information needs to be gathered when constructing or initializing TypeDeserializer instance.

Also: direct annotation access should be avoided: all access must be through `AnnotationIntrospector` (to allow for mix-in annotations, support for non-Jackson annotations (like JAXB))